File: tools/CliShim.UnitTests/CommandLineTests.cs
// Copyright (c) Microsoft Corporation
// The Microsoft Corporation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace PowerToys.CliShim.UnitTests;

[TestClass]
public sealed class CommandLineTests
{
    [DataTestMethod]

    // Normal shell launches: unquoted program name, ends at first whitespace.
    [DataRow("fancyzones arg", "arg")]
    [DataRow("fancyzones a b c", "a b c")]
    [DataRow("fancyzones", "")]
    [DataRow("filelocksmith", "")]

    // Quoted program name (path with spaces): ends at the closing quote, no backslash-escaping.
    [DataRow(@"""C:\Program Files\PowerToys\cli\fancyzones.exe"" arg", "arg")]
    [DataRow(@"""C:\Program Files\PowerToys\cli\fancyzones.exe""", "")]

    // The user's exact quoting in the tail is preserved verbatim (the whole point of the shim).
    [DataRow(@"""C:\cli\fancyzones.exe"" ""a b""", @"""a b""")]
    [DataRow(@"fancyzones --path ""C:\a b\c.png""", @"--path ""C:\a b\c.png""")]

    // Tabs count as whitespace for both the argv[0] terminator and the trim.
    [DataRow("fancyzones\targ", "arg")]
    [DataRow("fancyzones \t arg", "arg")]

    // Regression: a command line padded with leading whitespace must NOT leak the program name
    // (a non-shell parent can pass this via CreateProcessW; the OS loader never does).
    [DataRow("  fancyzones arg", "arg")]
    [DataRow(" fancyzones", "")]
    [DataRow(@"  ""C:\cli\fancyzones.exe"" arg", "arg")]

    // Degenerate inputs.
    [DataRow("", "")]

    // Unterminated argv[0] quote: ends at end-of-string (CRT-faithful), so no arguments remain.
    [DataRow(@"""C:\Program Files\app", "")]
    public void StripArgumentZero_ReturnsForwardedTail(string commandLine, string expected)
    {
        Assert.AreEqual(expected, CommandLine.StripArgumentZero(commandLine));
    }
}

File: tools/CliShim.UnitTests/CommandLineTests.cpp
// Copyright (c) Microsoft Corporation
// The Microsoft Corporation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

#include <string>

#include <CppUnitTest.h>

#include "CommandLine.h"

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace CliShimUnitTests
{
    TEST_CLASS(CommandLineTests)
    {
    public:
        TEST_METHOD(StripArgumentZero_ReturnsForwardedTail)
        {
            struct Case
            {
                const wchar_t* commandLine;
                const wchar_t* expected;
            };

            const Case cases[] = {
                { L"PowerToys.FancyZones.CLI arg", L"arg" },
                { L"PowerToys.FancyZones.CLI a b c", L"a b c" },
                { L"PowerToys.FancyZones.CLI", L"" },
                { L"PowerToys.FileLocksmith.CLI", L"" },

                { LR"("C:\Program Files\PowerToys\bin\PowerToys.FancyZones.CLI.exe" arg)", L"arg" },
                { LR"("C:\Program Files\PowerToys\bin\PowerToys.FancyZones.CLI.exe")", L"" },

                // Quotes only toggle, so a quoted argv[0] does not end at the closing quote: an
                // argument glued to it is part of the program name. CommandLineToArgvW would
                // forward `--help` here; the CRT - and so every target CLI - sees no arguments.
                { LR"("C:\bin\PowerToys.FancyZones.CLI.exe"--help)", L"" },

                // Partially quoted program names - the batch idiom of quoting only the variable,
                // `"%ProgramFiles%"\PowerToys\bin\...`, which cmd.exe passes through verbatim - are
                // resolved exactly as the CRT resolves them, so the target sees the same tail it
                // would have seen had the caller invoked it directly rather than through the shim.
                { LR"("C:\Program Files"\PowerToys\bin\PowerToys.FancyZones.CLI.exe arg)", L"arg" },
                { LR"(C:\Program" Files"\PowerToys\bin\PowerToys.FancyZones.CLI.exe arg)", L"arg" },
                { LR"("C:\Program Files"\PowerToys\bin\PowerToys.FancyZones.CLI.exe)", L"" },

                { LR"("C:\bin\PowerToys.FancyZones.CLI.exe" "a b")", LR"("a b")" },
                { LR"(PowerToys.FancyZones.CLI --path "C:\a b\c.png")", LR"(--path "C:\a b\c.png")" },

                { L"PowerToys.FancyZones.CLI\targ", L"arg" },
                { L"PowerToys.FancyZones.CLI \t arg", L"arg" },

                // Non-shell CreateProcessW callers can prepend whitespace; argv[0] must not leak.
                { L"  PowerToys.FancyZones.CLI arg", L"arg" },
                { L" PowerToys.FancyZones.CLI", L"" },
                { LR"(  "C:\bin\PowerToys.FancyZones.CLI.exe" arg)", L"arg" },

                { L"", L"" },

                // An unterminated argv[0] quote consumes the remaining command line.
                { LR"("C:\Program Files\app)", L"" },
            };

            for (const Case& testCase : cases)
            {
                const std::wstring actual = CommandLine::StripArgumentZero(testCase.commandLine);
                const std::wstring message = std::wstring(L"input: <") + testCase.commandLine + L">";
                Assert::AreEqual(std::wstring(testCase.expected), actual, message.c_str());
            }
        }
    };
}
